src/client/debugger/jupyter/kernelDebugAdapter.ts

```diff
@@ -170,8 +171,11 @@ export class KernelDebugAdapter implements DebugAdapter, IKernelDebugAdapter, ID
     }
 
     public disconnect() {
-        void this.session.customRequest('disconnect', { restart: false });
-        this.endSession.fire(this.session);
+        if (!this.disconected) {
```

Comment:
helps calling disconnect only once. The risk is that the message might fail.

Reply:
This doesn't help us understand why it is sent more than once or what the real correct time to send it is. We could check this in but I think it would deserve more investigation

Reply:
I digged, and the reason we sometimes call it twice is that the cell completing always calls disconnect, even after a restart or interrupt. Here's a summary of all the cases:

-finishing, only the cell status calls disconnect
-clicking stop, the controller disconnects + the cell status
-interrupt, the interrupt + the status change
-restart, the restart + the status change

If I move this check to the status change, we ensure to only call disconnect once.